test_bandwidth.py

```diff
@@ -0,0 +1,204 @@
+import os
+import sys
+import time
+import argparse
+from abc import abstractmethod
+
+import torch
+import torch.distributed as dist
+from typing import List
+from prettytable import PrettyTable
+os.environ['CUDA_VISIBLE_DEVICES'] = '1,2,3,4'
+    
+def setup(rank, world_size):
+    os.environ["MASTER_ADDR"] = os.getenv("MASTER_ADDR", "localhost")
+    os.environ["MASTER_PORT"] = os.getenv("MASTER_PORT", "12355")
+    os.environ["RANK"] = os.getenv("RANK", "0")
+    os.environ["WORLD_SIZE"] = os.getenv("WORLD_SIZE", "1")
+    dist.init_process_group(world_size=world_size, rank=rank,
+                            init_method="env://", backend="nccl")
+    torch.cuda.set_device(rank)
+    device_id = torch.cuda.current_device()
+    print("Process running on GPU:", device_id)
+
+"""    
+    BandWidth Computation: https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md
+
+    B: Bus bandWidth of the device
+    S: Elements transformed when execute one collective op
+    t: Time cost when execute one collective op 
+    n: Ranks number
+
+    AllReduce:
+        B = S/t * (2*(n-1)/n) = algbw * (2*(n-1)/n)
+    AllGather:
+        B = S/t * (n-1)/n = algbw * (n-1)/n
+    Broadcast:
+        B = S/t = algbw 
+    AllToAll:
+        B = S/t * (n-1)/n = algbw * (n-1)/n
+"""
+
+class CollectiveOps:
+    @abstractmethod
+    def __init__(self, world_size: int):
+        self.world_size = world_size
+
+    @abstractmethod
+    def __call__(self, tensor: torch.Tensor) -> None:
+        pass
+
+class AllReduce(CollectiveOps):
+    def __call__(self, tensor: torch.Tensor) -> None:
+        tensor = tensor.contiguous()
+        dist.all_reduce(tensor)
+    
+    def ranks_factor(self):
+        return 2 * (self.world_size - 1) / self.world_size
+
+
+class AllGather(CollectiveOps):
+    def __call__(self, tensor: torch.Tensor) -> None:
+        sub_tensor_list = list(torch.chunk(tensor, self.world_size, dim=0))
+        dist.all_gather(sub_tensor_list, sub_tensor_list[0]) 
+
+    def ranks_factor(self):
+        return (self.world_size - 1) / self.world_size 
+
+class Broadcast(CollectiveOps):
+    def __call__(self, tensor: torch.Tensor,src) -> None:
+        tensor = tensor.contiguous()
+        dist.broadcast(tensor,src)
+
+    def ranks_factor(self):
+        return 1
+
+class AllToAll(CollectiveOps):
+    def __call__(self, tensor: torch.Tensor) -> None:
+        input_tensor_list = list(torch.chunk(tensor, self.world_size, dim=0))
+        output_tensor_list = input_tensor_list[:]
+        dist.all_to_all(output_tensor_list, input_tensor_list)
+
+    def ranks_factor(self):
+        return (self.world_size - 1) / self.world_size 
+
+def avg_time(op, tensor:torch.tensor, iters):
```

Comment:
`torch.tensor` is a function, not a class.

Reply:
If you have type hinting for one arg, you should add hinting for other args as well.

Reply:
Ok, I'll remove it